# Portfolio Optimization

1. Understanding the word Portfolio
A portfolio, in finance, refer to a collection of investments you hold/ It's like a basket containing various assets like stocks, bons, real estate, or even cash. The goal of a portfolio is to grow your walth over time and achieve your financial goals.

2. Why is it importance?
There are many reason why we care about it:
- **Diversification**: Imagine putting all your eggs in one basket. if that basket falls, you lose everything!! A portfolio alls you to spread your investments across different asset classes. This reduces overall risk because even if one asset performs porrly, other might hold steady, minimizing losses.

- **Matching Risk to Goals**: Your risk tolerance (how comfortable you are with potential losses) is crucial. it helps you align your ivestments with your goals.

- **Growth Potential**: By carefully selecting and managing your portfolio, you can aim to grow your wealth over time. THis can helps you achieve financial goals like retirement, a down payment on a house, or funding your child's education.




# **Everything you should know before starting this project**

# **Modern Portfolio Theory (MPT) Optimizing for risk and return**

MPT, developed by Harry Markowitz. is a framework for building optimal portfolios. it focuses on achieving the following: 

- **Maximize expected return**: Aiming for the highest possible average return from your investment over time. 

- **Minize risk**: Reducing the uncertrainty associated with your portfolios's performance and potential losses.

MPT emapsize diversification as the key to achieving these goals. 

- **Efficient Frontier**: This is a theoretical concept representing the set of optimal portfolios that offer the highest expected return for given level of rish, or the lowest lowest level of risk for a given expected return. Portfolio on the efficient frontier provide the best possible combination of risk and return for a given investment strategy.


# **Sharpe Ration:Measuring Risk-Adjusted Return**

The Sharpe Ration is a metric used to evaluate the performance of an investment. It considers both the **average return** (how much your portfolio gains on average) and the **level of risk** (Volatility) involed.

Sharpe Ratio = (R<sub>p</sub> - R<sub>f</sub>) / σ<sub>p</sub>

Where:

- R<sub>p</sub> is the expected portfolio return.
- R<sub>f</sub> is the risk-free rate of return.
- σ<sub>p</sub> is the standard deviation of the portfolio's excess return.


**In the portfolio optimization analysis below, we will examine the stock performance of four prominent tech companies: AAPL (Apple), GOOG (Google), NFLX (Netflix), and MSFT (Microsoft) over the past five years. Our objective is to determine which company presents the most promising investment opportunity.👇👇👇**

In [1]:
# import python library
import pandas as pd
import numpy as np
import seaborn as sns
import datetime as datetime
import matplotlib.pyplot as plt
import plotly.graph_objects as go
plt.style.use('fivethirtyeight')

In [2]:
assets = pd.read_csv("/kaggle/input/stock-portfolio-in-the-market/Portfolio.csv")

"We'll focus solely on the 'Adj Close' price for our analysis moving forward!"

In [3]:
assets.head

<bound method NDFrame.head of            Date        NFLX        AAPL        MSFT        GOOG
0     4/22/2019  377.339996   49.225201  117.806213   62.442001
1     4/23/2019  381.890015   49.935188  119.405365   63.227501
2     4/24/2019  374.230011   49.858189  118.996056   62.799999
3     4/25/2019  368.329987   49.405697  122.936890   63.172501
4     4/26/2019  374.850006   49.169853  123.641281   63.609001
...         ...         ...         ...         ...         ...
1254  4/15/2024  607.150024  172.690002  413.640015  156.330002
1255  4/16/2024  617.520020  169.380005  414.579987  156.000000
1256  4/17/2024  613.690002  168.000000  411.839996  156.880005
1257  4/18/2024  610.559998  167.039993  404.269989  157.460007
1258  4/19/2024  555.039978  165.000000  399.119995  155.720001

[1259 rows x 5 columns]>

In [4]:
df = pd.DataFrame()

for stock in assets:
    df[stock] = assets[stock]
    
print(df.head())

        Date        NFLX       AAPL        MSFT       GOOG
0  4/22/2019  377.339996  49.225201  117.806213  62.442001
1  4/23/2019  381.890015  49.935188  119.405365  63.227501
2  4/24/2019  374.230011  49.858189  118.996056  62.799999
3  4/25/2019  368.329987  49.405697  122.936890  63.172501
4  4/26/2019  374.850006  49.169853  123.641281  63.609001


In [5]:
# Title for the plot
title = "Portfolio Adj. Close Prices History"

# Get the stocks
my_stocks = df

# Create the figure
fig = go.Figure()

# Add traces for each stock
for c in my_stocks.columns.values:
    fig.add_trace(go.Scatter(x=my_stocks.index, y=my_stocks[c], mode='lines', name=c, line=dict(width=2)))

# Update layout
fig.update_layout(
    title=title,
    xaxis_title="Date",
    yaxis_title="Adj Close USD ($)",
    legend=dict(orientation="h", yanchor="bottom", y=1.02, xanchor="right", x=1),
    showlegend=True,
    template="plotly_white"
)

# Show the plot
fig.show()


This code calculates the daily simple return for each stock in the DataFrame df using the pct_change() function in Pandas. It computes the percentage change in stock price from one day to the next and stores the results in the variable returns.

In [6]:
# Calculate the daily simple returns
returns = df.drop(columns=['Date']).pct_change()

# Display the resulting DataFrame
print(returns)


          NFLX      AAPL      MSFT      GOOG
0          NaN       NaN       NaN       NaN
1     0.012058  0.014423  0.013574  0.012580
2    -0.020058 -0.001542 -0.003428 -0.006761
3    -0.015766 -0.009076  0.033117  0.005932
4     0.017702 -0.004774  0.005730  0.006910
...        ...       ...       ...       ...
1254 -0.025175 -0.021864 -0.019578 -0.017966
1255  0.017080 -0.019167  0.002272 -0.002111
1256 -0.006202 -0.008147 -0.006609  0.005641
1257 -0.005100 -0.005714 -0.018381  0.003697
1258 -0.090933 -0.012213 -0.012739 -0.011050

[1259 rows x 4 columns]


The covariance matrix is multiplied by 252 to annualize it. Since there are approximately 252 trading days in a year, this adjustment scales up the daily covariance estimates to represent a yearly timeframe. It helps in comparing covariance values across different time periods, making them more interpretable and consistent.

In [7]:
# Create and show the annualized covariance matrix
cov_matrix_annual = returns.cov() * 252
cov_matrix_annual

NFLX      AAPL      MSFT      GOOG
NFLX  0.211821  0.067739  0.068216  0.067427
AAPL  0.067739  0.101393  0.072745  0.067966
MSFT  0.068216  0.072745  0.092437  0.072969
GOOG  0.067427  0.067966  0.072969  0.102607

The portfolio variance is a measure of the dispersion of returns of a portfolio. It quantifies the risk of the portfolio, indicating how much the actual returns of the portfolio deviate from the expected returns. 

The formula to calculate the portfolio variance for a portfolio with \( n \) assets, each with weight \( w_i \) and covariance \( \sigma_{ij} \) between assets \( i \) and \( j \), is:

$$
\text{Portfolio Variance} = \sum_{i=1}^{n} \sum_{j=1}^{n} w_i w_j \sigma_{ij}
$$

In vectorized form, it can be represented as:

$$
\text{Portfolio Variance} = \text{weights}^T \cdot \text{Covariance Matrix} \cdot \text{weights}
$$

Where:
- weights is a vector of asset weights,

- Covariance Matrix is the covariance matrix of returns for all assets in the portfolio.

- weights^T represents the transpose of the weight vector.

-  cdot  denotes the matrix multiplication operation.

This formula takes into account both the individual volatilities of the assets and their pairwise correlations, providing a comprehensive measure of portfolio risk.


In [8]:
#Initialize the weights
weights = np.array([0.25, 0.25, 0.25, 0.25])

# Calculate the portfolio variance
port_variance = np.dot(weights.T, np.dot(cov_matrix_annual, weights))
port_variance

0.08389898799374684

Portfolio volatility is a measure of the overall risk or variability of returns of a portfolio. It indicates the extent of fluctuation in the portfolio's value over time. The portfolio volatility formula calculates the standard deviation of the portfolio's returns, which represents the average deviation of returns from the mean return.

The formula for portfolio volatility is the square root of the portfolio variance:

$$
\text{Portfolio Volatility} = \sqrt{\text{Portfolio Variance}}
$$

Where:
- Portfolio Variance is the measure of dispersion of returns in the portfolio, accounting for both individual asset volatilities and their pairwise correlations.

Portfolio volatility provides investors with insights into the potential risk associated with holding a particular combination of assets. It's a crucial metric used in portfolio management to assess risk levels and make informed investment decisions.


In [9]:
# Calculate the portfolio volatility and standard diviation
port_volatility = np.sqrt(port_variance)
port_volatility

0.28965322023714296

The calculation of the annual portfolio return involves determining the average return generated by the portfolio over a specified time period, typically on an annual basis. This metric provides investors with an understanding of the expected return from holding the portfolio for a year.

The formula to calculate the annual portfolio return is:

$$
\text{Annual Portfolio Return} = \sum_{i=1}^{n} (\text{Mean Return of Asset } i \times \text{Weight of Asset } i) \times 252
$$

Where:
- n : is the number of assets in the portfolio,
- Mean Return of Asset:  is the average return of asset \( i \),
- Weight of Asset is the proportion of asset \( i \) in the portfolio,
- \( 252 \) represents the number of trading days in a year (typically used for daily returns).

The result is the expected annual return from the portfolio, taking into account the weighted contributions of each asset's average return. This metric assists investors in evaluating the potential profitability of the portfolio over a one-year period.


In [10]:
# Calculate the annual portfolio return 
portfolio_simple_annual_return = np.sum(returns.mean() * weights) * 252
portfolio_simple_annual_return

0.2513366714151853

In [11]:
# Round the calculated values to two decimal places
percent_var = "{:.2f}%".format(port_variance * 100)
percent_vola = "{:.2f}%".format(port_volatility * 100)
percent_ret = "{:.2f}%".format(portfolio_simple_annual_return * 100)

# Display the results
print("Expected annual return:", percent_ret)
print("Annual volatility risk:", percent_vola)
print("Annual Variance:", percent_var)


Expected annual return: 25.13%
Annual volatility risk: 28.97%
Annual Variance: 8.39%


In [12]:
from pypfopt.efficient_frontier import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns
from pypfopt import plotting
import cvxpy as cp


ModuleNotFoundError: No module named 'pypfopt'

In [ ]:
!pip install pypfopt



# **Portfolio Optimization**

In [ ]:
# Portfolio Optimization 
# Assuming 'df' is your DataFrame containing the date column and other data|


# Calculate the expected returns and annualized sample covariance matrix of asset returns
mu = expected_returns.mean_historical_return(df)
S = risk_models.sample_cov(df)

# Optimize for maximum Sharpe ratio
ef = EfficientFrontier(mu, S, weight_bounds=(None, None))  # Ensure weights are between 0 and 1
ef.add_constraint(lambda w: sum(w) == 1)  # Add a constraint to ensure the sum of weights is 1
weights = ef.max_sharpe()  # Maximize Sharpe ratio
cleaned_weights = ef.clean_weights()  # Clean the weights for better readability
print(cleaned_weights)
ef.portfolio_performance(verbose=True)  # Display portfolio performance metrics



1. **Asset Allocation**: The output shows the optimized weights for each asset in the portfolio. In this case, the portfolio is allocated 0% to 'NFLX', 38.87% to 'AAPL', 61.13% to 'MSFT', and 0% to 'GOOG'.

2. **Expected Annual Return**: The expected annual return of the portfolio is 27.6%. This indicates the anticipated percentage gain or loss in the portfolio's value over the course of one year, based on historical returns.

3. **Annual Volatility**: The annual volatility of the portfolio is 29.1%. Volatility measures the degree of variation of a trading price series over time. In this context, it represents the risk associated with the portfolio's returns.

4. **Sharpe Ratio**: The Sharpe Ratio is a measure of risk-adjusted return, calculated by dividing the portfolio's excess return over the risk-free rate by its standard deviation of returns. A higher Sharpe Ratio indicates a better risk-adjusted return. In this case, the Sharpe Ratio is 0.88.

5. **Tuple Representation**: The tuple `(0.2758351635369392, 0.2905723147796068, 0.88045264646422)` provides the same information as the previous three points but in a tuple format. It includes the expected annual return, annual volatility, and Sharpe Ratio, respectively.

Overall, this output provides valuable insights into the optimized asset allocation, expected return, risk, and risk-adjusted return of the portfolio based on historical data and optimization techniques.

In [ ]:
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices

latest_prices = get_latest_prices(df)
weights = cleaned_weights
da = DiscreteAllocation(weights, latest_prices, total_portfolio_value=50000)


allocation, leftover = da.lp_portfolio()
print("Discrete allocation: ", allocation)
print("Fund remaining: ${:.2f}".format(leftover))

To check the value per share for each asset in your portfolio, you can divide the total value of each asset by the number of shares allocated to it. Here's how you can calculate the value per share for each asset:

1. **AAPL Value Per Share**: Divide the total value of AAPL by the number of shares allocated to AAPL.
2. **MSFT Value Per Share**: Divide the total value of MSFT by the number of shares allocated to MSFT.
3. **GOOG Value Per Share**: Divide the total value of GOOG by the number of shares allocated to GOOG.

Let's perform these calculations based on the previous example:

- **AAPL Total Value**: $17,700
- **AAPL Shares**: 118
- **MSFT Total Value**: $15,200
- **MSFT Shares**: 76
- **GOOG Total Value**: $2500
- **GOOG Shares**: 1

Now, divide the total value of each asset by the number of shares allocated to it to find the value per share.

Here are the calculations for the value per share:

1. **AAPL Value Per Share**: \(\frac{\$17,700}{118} \approx \$150\)
2. **MSFT Value Per Share**: \(\frac{\$15,200}{76} \approx \$200\)
3. **GOOG Value Per Share**: \(\frac{\$2500}{1} = \$2500\)

So, the value per share for AAPL is approximately \$150, for MSFT is approximately \$200, and for GOOG is \$2500.

In [ ]:
# Calculate the expected returns and annualized sample covariance matrix of asset returns
mu = expected_returns.mean_historical_return(df)
S = risk_models.sample_cov(df)

# Instantiate a new EfficientFrontier object with unrestricted weight bounds
ef = EfficientFrontier(mu, S, weight_bounds=(None, None))

# Add a constraint to ensure the sum of weights is 1
ef.add_constraint(lambda w: sum(w) == 1)

# Plot the efficient frontier
plotting.plot_efficient_frontier(ef)

# Optimize for maximum Sharpe ratio
weights = ef.max_sharpe()

# Retrieve the optimized weights
cleaned_weights = ef.clean_weights()
ef.portfolio_performance(verbose=True)

In [ ]:
import numpy as np
import plotly.graph_objs as go
from pypfopt import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns

# Number of random portfolios to generate
num_portfolios = 1000

# Randomly generate asset weights for each portfolio
weights = np.random.rand(num_portfolios, len(mu))
weights = weights / weights.sum(axis=1, keepdims=True)  # Ensure weights sum up to 1 for each portfolio

# Calculate expected returns and volatility for each portfolio
portfolio_returns = np.dot(weights, mu)
portfolio_volatility = np.sqrt(np.diag(np.dot(np.dot(weights, S), weights.T)))

# Calculate Sharpe ratio for each portfolio
risk_free_rate = 0.02  # Example risk-free rate
sharpe_ratio = (portfolio_returns - risk_free_rate) / portfolio_volatility

# Find the portfolio with the maximum Sharpe ratio
max_sharpe_idx = np.argmax(sharpe_ratio)
max_sharpe_return = portfolio_returns[max_sharpe_idx]
max_sharpe_volatility = portfolio_volatility[max_sharpe_idx]

# Find the portfolio with the minimum volatility
min_volatility_idx = np.argmin(portfolio_volatility)
min_volatility_return = portfolio_returns[min_volatility_idx]
min_volatility_volatility = portfolio_volatility[min_volatility_idx]

# Set thresholds for Sharpe ratio to define safe, optimal, and risky portfolios
safe_threshold = 1.5
risky_threshold = 0.5

# Classify portfolios based on Sharpe ratio
portfolio_colors = []
for sr in sharpe_ratio:
    if sr >= safe_threshold:
        portfolio_colors.append('green')  # Safe
    elif sr >= risky_threshold:
        portfolio_colors.append('blue')   # Optimal
    else:
        portfolio_colors.append('red')    # Risky

# Construct the Efficient Frontier
ef = EfficientFrontier(mu, S)
ef.min_volatility()  # Minimize volatility for Efficient Frontier
ef_volatility, ef_return, _ = ef.portfolio_performance()  # Note: Only one value returned

# Plot the Efficient Frontier and random portfolios
fig = go.Figure()

# Add the Efficient Frontier line
fig.add_trace(go.Scatter(
    x=[ef_volatility],
    y=[ef_return],
    mode='markers',
    marker=dict(color='green', size=10),
    name='Efficient Frontier'
))

# Add the random portfolios
fig.add_trace(go.Scatter(
    x=portfolio_volatility,
    y=portfolio_returns,
    mode='markers',
    marker=dict(color=portfolio_colors, size=5),
    name='Random Portfolios'
))

# Add the portfolio with the maximum Sharpe ratio
fig.add_trace(go.Scatter(
    x=[max_sharpe_volatility],
    y=[max_sharpe_return],
    mode='markers',
    marker=dict(color='yellow', size=10),
    name='Max Sharpe Ratio Portfolio'
))

# Add the portfolio with the minimum volatility
fig.add_trace(go.Scatter(
    x=[min_volatility_volatility],
    y=[min_volatility_return],
    mode='markers',
    marker=dict(color='orange', size=10),
    name='Min Volatility Portfolio'
))

# Customize layout
fig.update_layout(
    title="Efficient Frontier with Max Sharpe Ratio Portfolio",
    xaxis_title="Volatility (Standard Deviation)",
    yaxis_title="Expected Return",
    legend=dict(
        x=0.7,
        y=0.2,
        bgcolor='rgba(255, 255, 255, 0)',
        bordercolor='rgba(255, 255, 255, 0)'
    )
)

# Show the plot
fig.show()


**Note**: Unfortunately, the section utilizing pypfopt for portfolio optimization cannot be demonstrated here due to restrictions on external package installations within the Kaggle environment. However, you can download the notebook file provided and run it in a different Python Integrated Development Environment (IDE) or platform where you can install external packages. This will allow you to explore the portfolio optimization process using pypfopt and gain a deeper understanding of the methodology. We apologize for any inconvenience caused and encourage you to continue your exploration offline.